doc/specs/#885 - Terminal Settings Model/Actions Addendum.md

```diff
@@ -213,9 +216,15 @@ Introducing a parent mechanism to `ActionMap` allows it to understand where a `C
 
 `ActionMap` queries will need to check their parent when they cannot find a matching `InternalActionID` in their `_ActionMap`.
 
-Since `NameMap` is generated upon request, we will need to pass a `std::set<InternalActionID>` as we generate the `NameMap` across each layer. This will ensure that each `Command` is only added to the `NameMap` once. We will start from the current layer and move up the inheritance tree to ensure that the current layer takes priority.
+Since `NameMap` is generated upon request, we will need to use a `std::set<InternalActionID>` as we generate the `NameMap`. This will ensure that each `Command` is only added to the `NameMap` once. The name map will be generated as follows:
+1. Get an accumulated list of `Command`s from our parents
+2. Iterate over the list...
+   - Update `NameMap` with any new `Command`s (tracked by the `std::set<InternalActionID>`)
 
-Nested commands will be saved to their own map since they do not have an `InternalActionID`. To ensure layering is accomplished correctly, we will need to start from the top-most parent and update `NameMap`. As we go down the inheritance tree, any conflicts are resolved by prioritizing the current layer (the child). This ensures that the current layer takes priority. 
+Nested commands will be saved to their own map since they do not have an `InternalActionID`.
+- During `ActionMap`'s population, we must ensure to resolve any conflicts immediately. This means that any new `Command`s that generate a name conflicting with a nested command will take priority (and we'll remove the nested comand from its own map). Conversely, if a new nested command conflicts with an existing standard `Command`, we can ignore it because our generation of `NameMap` will handle it.
+- When populating `NameMap`, we must first add all of the standard `Command`s. To ensure layering is accomplished correctly, we will need to start from the top-most parent and update `NameMap`. As we go down the inheritance tree, any conflicts are resolved by prioritizing the current layer (the child). This ensures that the current layer takes priority.
+- After adding all of the standard `Command`s to the `NameMap`, we can then register all of the nested commands. Since nested commands have no identifier other than a name, we cannot use the `InternalActionID` heuristic. However, as mentioned earlier, we are updating our internal nested command map as we add new actions. So when we are generating the name map, we can assume that all of these nested commands now have priority. Thus, we simply add all of these nested commands to the name map. Any conflicts are resolved in favor of th nested command.

```

Comment:
I'm still worried about this.

Consider that you have the following nested/repeated commands:

```
Split Pane...
   Windows PowerShell
```

```
New Tab...
   Windows PowerShell
```

**those have the same name**. They are very much different commands. We're not keying anything off their names, right? Well, except the Name Map which will immediately and horribly explode when you do that.

Reply:
I think there seems to be some confusion from my word choice. When I say "nested commands", I always mean the root. So in your example, there are 2 nested commands:
- "split pane..."
- "new tab..."

#### Part 1: `AddAction` the nested command

`AddAction` receives these commands and does the following:
- is it a nested command?
   - Yes! Insert (or assign) "<name>" to `_NestedCommands`
- Done! 

#### Part 2: `AddAction` a conflicting command

Later, let's say we get a new `Command x` who's name is "Split Pane...". We have two paths here:
1. `x` is a nested command:
   - update `_NestedCommands` such that "Split Pane..." maps to `x`
2. `x` is a standard command: 
   - _remove_ `Split Pane...` from `_NestedCommands`
   - register `x` with `_ActionMap`

NOTE: We don't care _what_ is inside of these nested commands. We just care about the root.

#### Part 3: `NameMap` generation with our parents

Now we get to part 3 of our story: dealing with parents. We don't really care about our parents until we have to generate the `NameMap`.

1. add all of the nested actions
   - Due to part 2, we can assume that we have a disjoint set of names between standard and nested actions. This means that all names in `_NestedCommands` are new to those from the standard actions.
   - If we start from our parents' nested actions, then work our way down to the current layer, we can create a `NameMap` filled with just nested commands. We don't know if any of them will persist, but that's ok, because then we move on to step 2...
2. add all of the standard/consolidated actions
   - ANY collisions are just straight-up assigned to these actions. The only collisions we can expect are the ones from our ancestors' nested commands intersecting with our standard actions. If that happens, we have the right to prioritize our standard actions.